fly.io/cdn-2024-01-26/default.vcl

```diff
@@ -0,0 +1,117 @@
+# https://blog.markvincze.com/how-to-gracefully-fall-back-to-cache-on-5xx-responses-with-varnish/
+sub vcl_backend_response {
+  # Objects within ttl are considered fresh.
+  set beresp.ttl = 60s;
+
+  # Objects within grace are considered stale.
+  # Serve stale content while refreshing in the background.
+  # 🤔 QUESTION: should we vary this based on backend health?
+  set beresp.grace = 24h;
+
+  if (beresp.status >= 500) {
+    # Don't cache a 5xx response
+    set beresp.uncacheable = true;
+
+    # If is_bgfetch is true, it means that we've found and returned the cached
+    # object to the client, and triggered an asynchoronus background update. In
+    # that case, since backend returned a 5xx, we have to abandon, otherwise
+    # the previously cached object would be erased from the cache (even if we
+    # set uncacheable to true).
+    if (bereq.is_bgfetch) {
+      return (abandon);
+    }
+  }
+
+  # 🤔 QUESTION: Should we configure beresp.keep?
+}
+
+# NOTE: vcl_recv is called at the beginning of a request, after the complete
+# request has been received and parsed. Its purpose is to decide whether or not
+# to serve the request, how to do it, and, if applicable, which backend to use.
+sub vcl_recv {
+  # https://varnish-cache.org/docs/7.4/users-guide/purging.html
+  if (req.method == "PURGE") {
+    return (purge);
+  }
+
+  # Implement a Varnish health-check
+  if (req.method == "GET" && req.url == "/varnish_status") {
+    return(synth(204));
+  }
+}
+
+# https://gist.github.com/leotsem/1246511/824cb9027a0a65d717c83e678850021dad84688d#file-default-vcl-pl
+# https://varnish-cache.org/docs/7.4/reference/vcl-var.html#obj
+sub vcl_deliver {
+  # What is the remaining TTL for this object?
+  set resp.http.x-ttl = obj.ttl;
+  # What is the max object staleness permitted?
+  set resp.http.x-grace = obj.grace;
+
+  # Did the response come from Varnish or from the backend?
+  if (obj.hits > 0) {
+    set resp.http.x-cache = "HIT";
+  } else {
+    set resp.http.x-cache = "MISS";
+  }
+
+  # Is this object stale?
+  if (obj.ttl < std.duration(integer=0)) {
+    set resp.http.x-cache = "STALE";
+  }
+
+  # How many times has this response been served from Varnish?
+  set resp.http.x-cache-hits = obj.hits;
+}
+
+# TODOS:
+# - ✅ Run in debug mode (locally)
+# - ✅ Connect directly to app - not Fly.io Proxy 🤦
+# - ✅ Serve stale content + background refresh
+#   - QUESTION: Should the app control this via Surrogate-Control? Should we remove this header?
+#   - EXPLORE: varnishstat
+#   - EXPLORE: varnishtop
+#   - EXPLORE: varnishncsa -c -F '%m %U %H %{x-cache}o %{x-cache-hits}o'
+# - ✅ Serve stale content on backend error
+#   - https://varnish-cache.org/docs/7.4/users-guide/vcl-grace.html#misbehaving-servers
+# - If the backend gets restarted (e.g. new deploy), backend remains sick in Varnish
+#   - https://info.varnish-software.com/blog/two-minute-tech-tuesdays-backend-health
+#   - EXPLORE: varnishlog -g raw -i backend_health
```

Comment:
@jamesarosen this is by far the most problematic part of this config. If the app gets re-deployed, the backend remains sick. This is what I am seeing in the varnish logs:

<img width="1012" alt="Screenshot 2024-06-20 at 07 09 18" src="https://github.com/thechangelog/changelog.com/assets/3342/62eadeed-46a3-44ce-8ac2-9ffee596eb7c">

My interpretation of the above is that Varnish does not try to re-establish the TCP connection to the backend once this goes away. WDYT?

Reply:
I have two hypotheses:

### Hypothesis One: It Takes Longer Than You Think

This VCL config in your probe:
```vcl
.interval = 5s;
.window = 10;
.threshold = 5;
```

Means that the backend will only be marked healthy why 5 (`threshold`) of the last 10 (`window`) requests have returned a `200 OK`. Once a backend becomes unhealthy, it will take up to 30s (`(5 + 1) * 5s`) to recover.

Try waiting longer. If that works, the solution is probably to do rolling restarts and use Varnish to load-balance. Given the way Fly's 6PN addressing works, I don't think you can be sure that `top1.nearest.of` wouldn't resolve to the same thing as any other given host. Thus, I think you need not 2 but 3 backends:

1. `top1.closest.of` -- the primary backend
2. `iad` (or any other fixed value) -- failover 1
3. `sjc` (or any other fixed value) -- failover 2

You would set up probes (healthchecks) for all three. Then you can check the health status in `vcl_recv`:

```vcl
set req.backend = primary;
if(!req.backend.healthy) {
     set req.backend = failover1;
}
if(!req.backend.healthy) {
     set req.backend = failover2;
}
```

### Hypothesis Two: You're Not Getting 200s

Is is possible that the `/health` endpoint is returning a 302 or 307? Perhaps it's redirecting from `http://` to `https://`? What do the origin logs say?

Reply:
If you have the varnish container running and can exec into it to run `varnishlog` then you could see what kind of responses Varnish is getting from your backends and this may help with troubleshooting any sick backends.

I cloned the repo and built the container on my local machine to have a running Varnish container to demonstrate the `varnishlog` command. I had to adjust the vcl and set the backend to target `pipedream.changelog.com` since it was complaining about not being able to resolve `top1.nearest.of.changelog-2024-01-12.internal` from my machine. Even though `pipedream.changelog.com` was only providing `503` responses to me, it still works to demonstrate the use of `varnishlog` just fine.

Once I had it running I just had a different shell fire off a request through it `curl -sv http://127.0.0.1:9000/` to get some log details.

```
varnish@2a59337f72cd:/etc/varnish$ varnishlog
*   << BeReq    >> 3         
-   Begin          bereq 2 fetch
-   VCL_use        boot
-   Timestamp      Start: 1719172226.180749 0.000000 0.000000
-   BereqMethod    GET
-   BereqURL       /
-   BereqProtocol  HTTP/1.1
-   BereqHeader    Host: 127.0.0.1:9000
-   BereqHeader    User-Agent: curl/8.8.0
-   BereqHeader    Accept: */*
-   BereqHeader    X-Forwarded-For: 192.168.65.1
-   BereqHeader    Via: 1.1 2a59337f72cd (Varnish/7.4)
-   BereqHeader    Accept-Encoding: gzip
-   BereqHeader    X-Varnish: 3
-   VCL_call       BACKEND_FETCH
-   VCL_return     fetch
-   Timestamp      Fetch: 1719172226.180792 0.000042 0.000042
-   FetchError     backend default: unhealthy
-   Timestamp      Beresp: 1719172226.180795 0.000046 0.000003
-   Timestamp      Error: 1719172226.180796 0.000046 0.000000
-   BerespProtocol HTTP/1.1
-   BerespStatus   503
-   BerespReason   Backend fetch failed
-   BerespHeader   Date: Sun, 23 Jun 2024 19:50:26 GMT
-   BerespHeader   Server: Varnish
-   VCL_call       BACKEND_ERROR
-   BerespHeader   Content-Type: text/html; charset=utf-8
-   BerespHeader   Retry-After: 5
-   VCL_return     deliver
-   Storage        malloc Transient
-   Length         278
-   BereqAcct      0 0 0 0 0 0
-   End            

*   << Request  >> 2         
-   Begin          req 1 rxreq
-   Timestamp      Start: 1719172226.180555 0.000000 0.000000
-   Timestamp      Req: 1719172226.180555 0.000000 0.000000
-   VCL_use        boot
-   ReqStart       192.168.65.1 57563 http
-   ReqMethod      GET
-   ReqURL         /
-   ReqProtocol    HTTP/1.1
-   ReqHeader      Host: 127.0.0.1:9000
-   ReqHeader      User-Agent: curl/8.8.0
-   ReqHeader      Accept: */*
-   ReqHeader      X-Forwarded-For: 192.168.65.1
-   ReqHeader      Via: 1.1 2a59337f72cd (Varnish/7.4)
-   VCL_call       RECV
-   VCL_return     hash
-   VCL_call       HASH
-   VCL_return     lookup
-   VCL_call       MISS
-   VCL_return     fetch
-   Link           bereq 3 fetch
-   Timestamp      Fetch: 1719172226.181146 0.000590 0.000590
-   RespProtocol   HTTP/1.1
-   RespStatus     503
-   RespReason     Backend fetch failed
-   RespHeader     Date: Sun, 23 Jun 2024 19:50:26 GMT
-   RespHeader     Server: Varnish
-   RespHeader     Content-Type: text/html; charset=utf-8
-   RespHeader     Retry-After: 5
-   RespHeader     X-Varnish: 2
-   RespHeader     Age: 0
-   RespHeader     Via: 1.1 2a59337f72cd (Varnish/7.4)
-   VCL_call       DELIVER
-   RespHeader     x-ttl: -0.000
-   RespHeader     x-grace: 0.000
-   RespHeader     x-cache: MISS
-   RespUnset      x-cache: MISS
-   RespHeader     x-cache: STALE
-   RespHeader     x-cache-hits: 0
-   VCL_return     deliver
-   Timestamp      Process: 1719172226.181170 0.000614 0.000024
-   Filters        
-   RespHeader     Content-Length: 278
-   RespHeader     Connection: keep-alive
-   Timestamp      Resp: 1719172226.181219 0.000663 0.000048
-   ReqAcct        77 0 77 315 278 593
-   End            

*   << Session  >> 1         
-   Begin          sess 0 HTTP/1
-   SessOpen       192.168.65.1 57563 http 172.17.0.2 9000 1719172226.180510 24
-   Link           req 2 rxreq
-   SessClose      REM_CLOSE 0.001
-   End            
```

The way this log works is you get multiple pieces, the **Session**, **BeReq**, and **Request**. It ends up listing this in the order that things finish and you have to hunt around a bit for where things started and figure out the chain of events. 

1. Here the **Session** has an identifier of `1` and is the starting point, from there you can see a reference to `Link` making a `req 2`. 
2. This is where the beginning of **Request** 2 starts and there is a reference that **Request** 2 was called by the id `1` in `Begin          req 1 rxreq`. You work through a few lines of the **Request** and you'll see another `Link` that points to `bereq 3 fetch` which is where it makes a backend request.
3. This is the point where **BeReq** starts and if you look up at **BeReq**, you'll see it contains a reference to being called from the identifier `2` in `Begin          bereq 2 fetch`. You can see the pieces of the request that **BeReq** received, where it makes a call to the backend, some timestamps of the various events and the details of the response it received from the backend, and then the **BeReq** finishes which is why it is first in the log entries.
4. Next you have the continuation of **Request** 2 where it receives a response and lists all the details of the response it got back from **BeReq** then it finishes
5.  Finally you are back to **Session** and it finishes

If you have a Varnish instance with a bunch of activity it's useful to use some command line options for `varnishlog` to filter things a bit to only show the log events you want, and a nice reference for the command line options is here: https://www.varnish-software.com/developers/tutorials/varnishlog/

I had to go [reference an old gist I made](https://gist.github.com/mttjohnson/da3b048282d49e6ff356) when doing this more regularly. Using the `-q` option is useful to filter the logs and only show logs coming from your own IP.
```
varnishlog -q 'ReqHeader:X-Forwarded-For eq "192.168.65.1" or BereqHeader:X-Forwarded-For eq "192.168.65.1"'
```
...this would only end up showing the Request and BeReq events from the log since the Session wouldn't have these headers

```
varnishlog -q 'ReqURL eq "/friends/50" or BereqURL eq "/friends/50"'
```
...this would only show Request and BeReq events to a specific URL.

Reply:
I have another hypothesis: Fly is assigning a new IP address during the deploy. If you give a DNS name for a `.host`, [Varnish only resolves that once, at launch](https://medium.com/@archonkulis/varnish-resolves-backend-ip-only-at-launch-time-64b4c103daed). If Fly assigns a new IP, you'll be hammering someone else's service with healthcheck requests 😓

Reply:
I recommend upvoting [this Fly.io feature request](https://community.fly.io/t/feature-request-configurable-private-dns-srv-records/8201) to support SRV records. [libvmod](https://github.com/nigoroll/libvmod-dynamic) works with [getdns](https://getdnsapi.net/) to support `SRV` records. That would mean you wouldn't have to specify the port here. It would pick the port up from DNS.